packages/components/src/services/hds-time.ts

```diff
@@ -0,0 +1,280 @@
+
+      // If the difference in date is greater than the threshold for showing the
+      // relative time then switch the display key.
+      if (difference.absValueInMs > THRESHOLD_RELATIVE_TIME_IN_MS) {
+        displayKey = HdsDisplayKeyValues.FriendlyLocal;
+      }
+    }
+
+    const options: DefaultDisplayType | undefined =
+      DEFAULT_DISPLAY_MAPPING[displayKey];
+
+    return {
+      options,
+      difference,
+      relative: this.selectTimeRelativeUnit(difference),
+    };
+  }
+
+  // Formats the value of a relative unit.
+  formatTimeRelativeUnit(
+    value: number,
+    unit: string
+  ): { value: number; unit: string } {
+    return {
+      value: Math.trunc(value),
+      unit,
+    };
+  }
+
+  // Selects an appropriate display format for the difference.
+  selectTimeRelativeUnit(
+    { absValueInMs, valueInMs }: { absValueInMs: number; valueInMs: number },
+    thresholds: {
+      second: number;
+      minute: number;
+      hour: number;
+      day: number;
+    } = DEFAULT_RELATIVE_THRESHOLDS
+  ): { value: number; unit: string } {
+    if (absValueInMs < thresholds[HdsTimeRelativeUnitValues.Second]) {
+      return this.formatTimeRelativeUnit(
+        valueInMs / SECOND_IN_MS,
+        HdsTimeRelativeUnitValues.Second
+      );
+    }
+
+    if (absValueInMs < thresholds[HdsTimeRelativeUnitValues.Minute]) {
+      return this.formatTimeRelativeUnit(
+        valueInMs / MINUTE_IN_MS,
+        HdsTimeRelativeUnitValues.Minute
+      );
+    }
+
+    if (absValueInMs < thresholds[HdsTimeRelativeUnitValues.Hour]) {
+      return this.formatTimeRelativeUnit(
+        valueInMs / HOUR_IN_MS,
+        HdsTimeRelativeUnitValues.Hour
+      );
+    }
+
+    if (absValueInMs < thresholds[HdsTimeRelativeUnitValues.Day]) {
+      return this.formatTimeRelativeUnit(
+        valueInMs / DAY_IN_MS,
+        HdsTimeRelativeUnitValues.Day
+      );
+    }
+
+    return this.formatTimeRelativeUnit(
+      valueInMs / WEEK_IN_MS,
+      HdsTimeRelativeUnitValues.Week
+    );
+  }
+
+  // Gets the currently subscribed listeners.
+  timeDifference(
+    startDate: number,
+    endDate: Date
+  ): { absValueInMs: number; valueInMs: number } {
+    const valueInMs = Number(endDate) - Number(startDate);
+    return {
+      absValueInMs: Math.abs(valueInMs),
+      valueInMs,
+    };
+  }
+
+  // Subscribes a listener to the ticking task for time changes.
+  register(id: Date): () => void {
+    this.#listeners.add(id);
+    // @ts-expect-error - TS2339: Property 'perform' does not exist on type '() => TaskGenerator<string | undefined>'
+    // note: we could potentially use taskFor via `ember-concurrency-ts` to avoid this exception
+    this.start.perform();
+    return (): void => {
+      this.unregister(id);
+    };
+  }
+
+  // Unregisters listener for the time task.
+  unregister(id: Date): boolean {
+    return this.#listeners.delete(id);
+  }
+
+  @task({ drop: true })
+  *start(): TaskGenerator<string | undefined> {
```

Comment:
Question: What does the `*` before start mean?

Reply:
Woah, very nifty! No comments on if theres a better way to do it, I just haven't seen this before!